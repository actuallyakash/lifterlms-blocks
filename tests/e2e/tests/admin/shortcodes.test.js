/**
 * Test shortcode inserter
 *
 * @since [version]
 * @version [version]
 */

import {
	clickBlockAppender,
	getEditedPostContent,
} from '@wordpress/e2e-test-utils';

import {
	click,
	clickElementByText,
	createPost,
} from '@lifterlms/llms-e2e-test-utils';

// CSS selectors.
const TOOLBAR_SELECTOR = '.block-editor-block-contextual-toolbar',
	MODAL_SELCTOR = '.llms-shortcodes-modal';

/**
 * Retrieve a list of the titles of all shortcodes in the inserter table.
 *
 * @since [version]
 *
 * @return {string[]} List of shortcode titles.
 */
async function getVisibleTitles() {
	return await page.$$eval( `${ MODAL_SELCTOR } .llms-table tbody tr td:first-child`, els => els.map( ( { textContent } ) => textContent ) );
}

/**
 * Retrieve a list of the shortcodes visible in the inserter table.
 *
 * @since [version]
 *
 * @return {string[]} List of shortcodes.
 */
async function getVisibleCodes() {
	return await page.$$eval( `${ MODAL_SELCTOR } .llms-table tbody tr td:nth-child(2)`, els => els.map( ( { textContent } ) => textContent ) );
}

/**
 * Clear the currently selected input
 *
 * @since [version]
 *
 * @param {Integer} length Number of characters to remove.
 * @return {void}
 */
async function clearSearch( length ) {

	for ( let i = 0; i < length; i++ ) {
		await page.keyboard.press('Backspace');
	}

}

/**
 * Filter the shortcode list by search query
 *
 * @since [version]
 *
 * @param {string} query Search query.
 * @return {void}
 */
async function filterList( query ) {
	await clickElementByText( 'Filter by label', `${ MODAL_SELCTOR } label` );
	await page.keyboard.type( query );
}

/**
 * Add a default value to the visible shortcodes
 *
 * @since [version]
 *
 * @param {string} val Default value.
 * @return {void}
 */
async function addDefaultValue( val ) {
	await clickElementByText( 'Default value', `${ MODAL_SELCTOR } label` );
	await page.keyboard.type( val );
}

/**
 * Close the inserter modal
 *
 * @since [version]
 *
 * @return {void}
 */
async function closeModal() {
	await click( `${ MODAL_SELCTOR } button[aria-label="Close dialog"]` );
}

describe( 'Admin/Shortcodes', () => {

	beforeAll( async() => {

		await createPost( 'post', 'Shortcodes' );

		await clickBlockAppender();

		await page.keyboard.type( 'Ipsum' );
		await page.keyboard.press( 'Tab' );

	} );

	beforeEach( async () => {

		await page.waitForSelector( TOOLBAR_SELECTOR );

		await click( `${ TOOLBAR_SELECTOR } button.components-dropdown-menu__toggle[aria-label="More"]` );
		await clickElementByText( 'Shortcodes', '.components-dropdown-menu__menu-item' );

		await page.waitForSelector( MODAL_SELCTOR );

	} );

	it ( 'should filter the list', async() => {

		const queries = [ 'name', 'user', 'email' ];

		for ( let i = 0; i < queries.length; i++ ) {

			const query = queries[ i ];

			await filterList( query );
			expect( await getVisibleTitles() ).toMatchSnapshot();
			await clearSearch( query.length );
			expect( await getVisibleTitles() ).toMatchSnapshot();

		}

		await closeModal();

	} );

	it ( 'should add a default value', async() => {

		await addDefaultValue( 'default' );
		expect( await getVisibleCodes() ).toMatchSnapshot();

		await clearSearch( 7 );
		expect( await getVisibleCodes() ).toMatchSnapshot();

		await closeModal();

	} );


	it ( 'should insert the shortcode', async() => {

		await filterList( 'display' );
		await click( `${ MODAL_SELCTOR } .llms-table tbody tr:first-child td:nth-child(3) button` );

		expect( await getEditedPostContent() ).toMatchSnapshot();

	} );

	it ( 'should insert the shortcode with a default value', async() => {

		await filterList( 'first' );
		await addDefaultValue( 'friend' );
		await click( `${ MODAL_SELCTOR } .llms-table tbody tr:first-child td:nth-child(3) button` );

		expect( await getEditedPostContent() ).toMatchSnapshot();

	} );


	it ( 'should allow custom shortcodes to be used', async() => {

		await filterList( 'fake' );
		await addDefaultValue( 'test' );

		// Shows the warning message.
		expect( await page.$eval( `${ MODAL_SELCTOR } .llms-error`, ( { textContent }) => textContent ) ).toMatchSnapshot();

		// Added to the list.
		expect( await getVisibleTitles() ).toMatchSnapshot();
		expect( await getVisibleCodes() ).toMatchSnapshot();

		// Inserted.
		await click( `${ MODAL_SELCTOR } .llms-table tbody tr:first-child td:nth-child(3) button` );
		expect( await getEditedPostContent() ).toMatchSnapshot();

	} );

} );
