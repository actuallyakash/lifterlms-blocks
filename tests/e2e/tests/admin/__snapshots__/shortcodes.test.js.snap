// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Admin/Shortcodes should add a default value 1`] = `
Array [
  "[llms-user user_login or=\\"default\\"]",
  "[llms-user user_email or=\\"default\\"]",
  "[llms-user first_name or=\\"default\\"]",
  "[llms-user last_name or=\\"default\\"]",
  "[llms-user display_name or=\\"default\\"]",
  "[llms-user llms_billing_address_1 or=\\"default\\"]",
  "[llms-user llms_billing_address_2 or=\\"default\\"]",
  "[llms-user llms_billing_city or=\\"default\\"]",
  "[llms-user llms_billing_country or=\\"default\\"]",
  "[llms-user llms_billing_state or=\\"default\\"]",
  "[llms-user llms_billing_zip or=\\"default\\"]",
  "[llms-user llms_phone or=\\"default\\"]",
]
`;

exports[`Admin/Shortcodes should add a default value 2`] = `
Array [
  "[llms-user user_login]",
  "[llms-user user_email]",
  "[llms-user first_name]",
  "[llms-user last_name]",
  "[llms-user display_name]",
  "[llms-user llms_billing_address_1]",
  "[llms-user llms_billing_address_2]",
  "[llms-user llms_billing_city]",
  "[llms-user llms_billing_country]",
  "[llms-user llms_billing_state]",
  "[llms-user llms_billing_zip]",
  "[llms-user llms_phone]",
]
`;

exports[`Admin/Shortcodes should allow custom shortcodes to be used 1`] = `"No fields found matching your search but you can use the shortcode below if the meta information exists in the database."`;

exports[`Admin/Shortcodes should allow custom shortcodes to be used 2`] = `
Array [
  "Custom User Information",
]
`;

exports[`Admin/Shortcodes should allow custom shortcodes to be used 3`] = `
Array [
  "[llms-user fake or=\\"test\\"]",
]
`;

exports[`Admin/Shortcodes should allow custom shortcodes to be used 4`] = `
"<!-- wp:paragraph -->
<p>Ipsum<span class=\\"llms-user-sc-wrap\\">[llms-user display_name][llms-user first_name or=\\"friend\\"][llms-user fake or=\\"test\\"]</span></p>
<!-- /wp:paragraph -->"
`;

exports[`Admin/Shortcodes should filter the list 1`] = `
Array [
  "Username",
  "First Name",
  "Last Name",
  "Display Name",
]
`;

exports[`Admin/Shortcodes should filter the list 2`] = `
Array [
  "Username",
  "Email Address",
  "First Name",
  "Last Name",
  "Display Name",
  "Address",
  "",
  "City",
  "Country",
  "State / Region",
  "Postal / Zip Code",
  "Phone Number",
]
`;

exports[`Admin/Shortcodes should filter the list 3`] = `
Array [
  "Username",
  "Email Address",
]
`;

exports[`Admin/Shortcodes should filter the list 4`] = `
Array [
  "Username",
  "Email Address",
  "First Name",
  "Last Name",
  "Display Name",
  "Address",
  "",
  "City",
  "Country",
  "State / Region",
  "Postal / Zip Code",
  "Phone Number",
]
`;

exports[`Admin/Shortcodes should filter the list 5`] = `
Array [
  "Email Address",
]
`;

exports[`Admin/Shortcodes should filter the list 6`] = `
Array [
  "Username",
  "Email Address",
  "First Name",
  "Last Name",
  "Display Name",
  "Address",
  "",
  "City",
  "Country",
  "State / Region",
  "Postal / Zip Code",
  "Phone Number",
]
`;

exports[`Admin/Shortcodes should insert the shortcode 1`] = `
"<!-- wp:paragraph -->
<p>Ipsum<span class=\\"llms-user-sc-wrap\\">[llms-user display_name]</span></p>
<!-- /wp:paragraph -->"
`;

exports[`Admin/Shortcodes should insert the shortcode with a default value 1`] = `
"<!-- wp:paragraph -->
<p>Ipsum<span class=\\"llms-user-sc-wrap\\">[llms-user display_name][llms-user first_name or=\\"friend\\"]</span></p>
<!-- /wp:paragraph -->"
`;
